### PR TITLE
Add float16 approve for multi-type part

### PR DESCRIPTION
**Pull request by Writer-X**
issue：https://github.com/milvus-io/milvus/issues/22837

Add bfloat16 vector, add the index part of float16 vector.

### CODE REVIEW

File: internal/core/src/query/SearchBruteForce.cpp [hunk 1]
```diff
@@ -101,14 +72,26 @@ BruteForceSearch(const dataset::SearchDataset& dataset,
...
+            res = knowhere::BruteForce::RangeSearch<uint8_t>(
+                base_dataset, query_dataset, config, bitset);
+        }
         milvus::tracer::AddEvent("knowhere_finish_BruteForce_RangeSearch");
         if (!res.has_value()) {
             PanicInfo(KnowhereError,
-                      "failed to range search: {}: {}",
-                      KnowhereStatusString(res.error()),
-                      res.what());
+                      fmt::format("failed to range search: {}: {}",
```
Comment (yah01) [inline internal/core/src/query/SearchBruteForce.cpp:92]: The `PanicInfo` supports to format the arguments, no need to call `fmt::format`, maybe improve this next PR

File: internal/proxy/task.go [hunk 2]
```diff
@@ -1461,7 +1461,7 @@ func (t *loadCollectionTask) Execute(ctx context.Context) (err error) {
 	for _, index := range indexResponse.IndexInfos {
 		fieldIndexIDs[index.FieldID] = index.IndexID
 		for _, field := range collSchema.Fields {
-			if index.FieldID == field.FieldID && (field.DataType == schemapb.DataType_FloatVector || field.DataType == schemapb.DataType_BinaryVector || field.DataType == schemapb.DataType_Float16Vector) {
+			if index.FieldID == field.FieldID && (field.DataType == schemapb.DataType_FloatVector || field.DataType == schemapb.DataType_BinaryVector || field.DataType == schemapb.DataType_Float16Vector || field.DataType == schemapb.DataType_BFloat16Vector) {
```
Comment (yah01) [inline internal/proxy/task.go:1464]: we need a util like `IsVectorType`

**Review by yah01 — COMMENTED** [review]
/lgtm

File: internal/core/src/query/SearchBruteForce.cpp (hunk 1 above)
Comment (Writer-X) [reply internal/core/src/query/SearchBruteForce.cpp:92]: Get.

File: internal/proxy/task.go (hunk 2 above)
Comment (Writer-X) [reply internal/proxy/task.go:1464]: Get.

**Comment by cqy123456** [comment]
rerun ut

**Review by sunby — APPROVED** [review]
/lgtm

**Comment by Writer-X** [comment]
rerun ut

File: internal/proxy/task_query.go [hunk 3]
```diff
@@ -74,7 +74,7 @@ func translateToOutputFieldIDs(outputFields []string, schema *schemapb.Collectio
 	outputFieldIDs := make([]UniqueID, 0, len(outputFields)+1)
 	if len(outputFields) == 0 {
 		for _, field := range schema.Fields {
-			if field.FieldID >= common.StartOfUserFieldID && field.DataType != schemapb.DataType_FloatVector && field.DataType != schemapb.DataType_BinaryVector && field.DataType != schemapb.DataType_Float16Vector {
+			if field.FieldID >= common.StartOfUserFieldID && field.DataType != schemapb.DataType_FloatVector && field.DataType != schemapb.DataType_BinaryVector && field.DataType != schemapb.DataType_Float16Vector && field.DataType != schemapb.DataType_BFloat16Vector {
```
Comment (czs007) [inline internal/proxy/task_query.go:77]: !isVectorType(field.DataType)?

File: internal/proxy/task_test.go [hunk 4]
```diff
@@ -173,7 +174,7 @@ func constructCollectionSchemaByDataType(collectionName string, fieldName2DataTy
 			Name:     fieldName,
 			DataType: dataType,
 		}
-		if dataType == schemapb.DataType_FloatVector || dataType == schemapb.DataType_BinaryVector || dataType == schemapb.DataType_Float16Vector {
+		if dataType == schemapb.DataType_FloatVector || dataType == schemapb.DataType_BinaryVector || dataType == schemapb.DataType_Float16Vector || dataType == schemapb.DataType_BFloat16Vector {
```
Comment (czs007) [inline internal/proxy/task_test.go:177]: isVectorType

File: tests/integration/util_query.go [hunk 5]
```diff
@@ -192,6 +192,17 @@ func constructPlaceholderGroup(nq, dim int, vectorType schemapb.DataType) *commo
 			}
 			values = append(values, ret)
 		}
+	// case schemapb.DataType_BFloat16Vector:
+	// 	placeholderType = commonpb.PlaceholderType_BFloat16Vector
+	// 	for i := 0; i < nq; i++ {
+	// 		total := dim * 2
+	// 		ret := make([]byte, total)
+	// 		_, err := rand.Read(ret)
+	// 		if err != nil {
```
Comment (czs007) [inline tests/integration/util_query.go:201]: just remove this

File: tests/integration/util_insert.go [hunk 6]
```diff
@@ -194,6 +209,16 @@ func GenerateFloat16Vectors(numRows, dim int) []byte {
 	return ret
 }
 
+// func GenerateBFloat16Vectors(numRows, dim int) []byte {
```
Comment (czs007) [inline tests/integration/util_insert.go:212]: just remove this

File: pkg/util/indexparamcheck/hnsw_checker.go [hunk 7]
```diff
@@ -31,7 +31,7 @@ func (c hnswChecker) CheckTrain(params map[string]string) error {
 }
 
 func (c hnswChecker) CheckValidDataType(dType schemapb.DataType) error {
-	if dType != schemapb.DataType_FloatVector && dType != schemapb.DataType_BinaryVector && dType != schemapb.DataType_Float16Vector {
+	if dType != schemapb.DataType_FloatVector && dType != schemapb.DataType_BinaryVector && dType != schemapb.DataType_Float16Vector && dType != schemapb.DataType_BFloat16Vector {
```
Comment (czs007) [inline pkg/util/indexparamcheck/hnsw_checker.go:34]: !IsVectorType(dType)

File: internal/core/thirdparty/knowhere/CMakeLists.txt [hunk 8]
```diff
@@ -12,8 +12,8 @@
 #-------------------------------------------------------------------------------
 
 # Update KNOWHERE_VERSION for the first occurrence
-set( KNOWHERE_VERSION 20a6289 )
-set( GIT_REPOSITORY  "https://github.com/zilliztech/knowhere.git")
+set( KNOWHERE_VERSION template_data )
+set( GIT_REPOSITORY  "https://github.com/cqy123456/zilliztech-knowhere.git")
```
Comment (czs007) [inline internal/core/thirdparty/knowhere/CMakeLists.txt:16]: Once the pull request from knowhere is merged, please ensure that the GIT_REPOSITORY is redirected to the right repository

**Comment by czs007** [comment]
Please perform a rebase on the master branch to address the previously encountered issue of failed UT compilation.

File: pkg/util/indexparamcheck/hnsw_checker.go (hunk 7 above)
Comment (Writer-X) [reply pkg/util/indexparamcheck/hnsw_checker.go:34]: In fact, the IsVectorType() function is defined in "internal/proxy/utils.go". Maybe I should find an appropriate location in "pkg/util" and define it again.

**Pull request merged**